### PR TITLE
sql/catalog: add region_liveness table

### PR DESCRIPTION
**Pull request by fqazi**
This adds a new region_liveness table and a migration to create this table for existing clusters, allowing us to introduce the concept of region liveness in later PRs. This will allow certain system tables to be configured  as SURVIVE REGION later on.

Epic: [CRDB-28158](https://cockroachlabs.atlassian.net/browse/CRDB-28158)

### CODE REVIEW

File: pkg/sql/region_util.go [hunk 1]
```diff
@@ -2441,6 +2441,7 @@ func (p *planner) optimizeSystemDatabase(ctx context.Context) error {
 		"namespace",
 		"table_statistics",
 		"web_sessions",
+		"region_liveness",
```
Comment (JeffSwenson) [inline pkg/sql/region_util.go:2444]: What do you think of transmuting the crdb_region column into the proper enum type? The loop that processes the regional by row tables contains an example of how to change the type.

We would need to use the same logic in the migration to adjust the column type when creating the table.

**Review by jeffswenson — COMMENTED** [review]
This is great 💯

File: pkg/upgrade/upgrades/create_region_liveness.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+			if err != nil {
+				return err
+			}
+
+			column := catalog.FindColumnByName(tbl, "crdb_region")
+			if column == nil {
+				return errors.AssertionFailedf("unable to find region column")
+			}
+
+			enumTypeDesc, err := txn.Descriptors().MutableByID(txn.KV()).Type(ctx, regionEnumID)
```
Comment (JeffSwenson) [inline pkg/upgrade/upgrades/create_region_liveness.go:84]: The migration descriptor transaction appears to be missing the logic to set the back reference on the `enumTypeDesc`.

Comment (fqazi) [reply pkg/upgrade/upgrades/create_region_liveness.go:82]: Good point, also fixed some other stuff for multi-tenant which was bad.

**Review by jeffswenson — APPROVED** [review]
LGTM

**Pull request merged**